### PR TITLE
CORTX-30314:Implement REST API for tagging node as a permanent failure

### PR DESCRIPTION
**Pull request by Pranali04796**
Signed-off-by: Pranali04796 <pranali.ugale@seagate.com>

# Problem Statement
- Jira: [CORTX-30314](https://jts.seagate.com/browse/CORTX-30314)
- Implement REST API for tagging node as a permanent failure

# Design
-  [Node failure API specifications](https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/983400463/Node+failure+API+specifications)
-  [Action Status Update LLD](https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/891781277/Action+Status+Update+LLD#Connection)
# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM
- Build details:

- Deployment details:

- Test case1: Happy path scenario.

- Test case2: Invalid resource.

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: csm/common/ha/cluster_management/operations_factory.py [hunk 1]
```diff
@@ -61,7 +61,8 @@ def get_operation(self, operation: str) -> operations.Operation:
             nodeOperation = operations.NodeStopOperation()
         elif operation == "poweroff":
             nodeOperation = operations.NodePoweroffOperation()
-
+        elif operation == "mark_node_failure":
```
Comment (rohitkolapkar) [inline csm/common/ha/cluster_management/operations_factory.py:64]: read this from const -> MARK_NODE_FAILURE

File: csm/conf/post_install.py [hunk 2]
```diff
@@ -123,3 +126,28 @@ def create(self):
...
+        node_ids = []
+        Log.info("Fetching node ids from each storage set.")
+        num_storage_sets = Conf.get(const.CONSUMER_INDEX, const.NUM_STORAGE_SET)
+        for index in range(num_storage_sets):
+            node_ids =  node_ids + Conf.get(const.CONSUMER_INDEX,
+                                            f'cluster>storage_set[{index}]>nodes')
+        Log.info("Setting node id.")
+        Conf.set(const.CSM_GLOBAL_INDEX, const.NUM_NODE_ID, str(len(node_ids)))
+        for index, node_id in enumerate(node_ids):
+            Conf.set(const.CSM_GLOBAL_INDEX, f"{const.KEY_NODE_ID}[{index}]", str(node_id))
```
Comment (rohitkolapkar) [inline csm/conf/post_install.py:153]: Add a blank line at the end of the file.

File: csm/common/ha/cluster_management/operations.py [hunk 3]
```diff
@@ -179,3 +181,56 @@ def execute(self, cluster_manager, **kwargs):
...
+        Create payload for tagging node failure.
+        :param : arguments with node_id
+        :returns:
+        """
+        cluster_id = Conf.get(const.CSM_GLOBAL_INDEX, const.KEY_CLUSTERID)
+        arguments = kwargs.get('arguments', "")
+        node_id = arguments['id']
+        health_attrs = {
+            f'{HealthAttr.SOURCE}': 'csm',
+            f'{HealthAttr.CLUSTER_ID}': cluster_id ,
```
Comment (rohitkolapkar) [inline csm/common/ha/cluster_management/operations.py:214]: read 'csm' from const file, check if it is already present.

File: csm/common/ha/cluster_management/operations.py [hunk 4]
```diff
@@ -179,3 +181,56 @@ def execute(self, cluster_manager, **kwargs):
...
+            f'{HealthAttr.SITE_ID}': const.NOT_DEFINED,
+            f'{HealthAttr.RACK_ID}': const.NOT_DEFINED,
+            f'{HealthAttr.STORAGESET_ID}': const.NOT_DEFINED,
+            f'{HealthAttr.NODE_ID}': node_id,
+            f'{HealthAttr.RESOURCE_TYPE}':const.NODE,
+            f'{HealthAttr.RESOURCE_ID}': node_id,
+            f'{HealthAttr.RESOURCE_STATUS}': 'failed',
```
Comment (rohitkolapkar) [inline csm/common/ha/cluster_management/operations.py:221]: read 'failed' from const file.

**Review by rohitkolapkar — CHANGES_REQUESTED** [review]

File: csm/common/ha/cluster_management/operations_factory.py (hunk 1 above)
Comment (Pranali04796) [reply csm/common/ha/cluster_management/operations_factory.py]: Addressed

File: csm/conf/post_install.py (hunk 2 above)
Comment (Pranali04796) [reply csm/conf/post_install.py]: Addressed

File: csm/common/ha/cluster_management/operations.py [hunk 5]
```diff
@@ -179,3 +181,56 @@ def execute(self, cluster_manager, **kwargs):
...
+        raise InvalidRequest('Request body is missing or invalid request body.')
+
+    def _create_payload(self, **kwargs):
+        """
+        Create payload for tagging node failure.
+        :param : arguments with node_id
+        :returns:
+        """
+        cluster_id = Conf.get(const.CSM_GLOBAL_INDEX, const.KEY_CLUSTERID)
+        arguments = kwargs.get('arguments', "")
```
Comment (abhijit1patil) [inline csm/common/ha/cluster_management/operations.py:210]: Keep 'arguments'/'id' in Const file if not present

Comment (Pranali04796) [reply csm/common/ha/cluster_management/operations.py:210]: Updated

File: csm/common/ha/cluster_management/operations.py [hunk 6]
```diff
@@ -21,6 +21,8 @@
 from csm.core.blogic import const
 from csm.common.errors import InvalidRequest, CsmInternalError
 from cortx.utils.log import Log
+from cortx.utils.conf_store.conf_store import Conf
+from cortx.utils.event_framework.health import HealthEvent, HealthAttr
```
Comment (pawankumarsrivastava) [inline csm/common/ha/cluster_management/operations.py:25]: Remove health schema. 
Change endpoint and remove 'management' from it.

Comment (pawankumarsrivastava) [inline csm/common/ha/cluster_management/operations.py:25]: Remove health schema. 
Change endpoint and remove 'management' from it.

File: csm/common/ha/cluster_management/operations.py [hunk 7]
```diff
@@ -179,3 +181,56 @@ def execute(self, cluster_manager, **kwargs):
             err_msg = f"{const.CLUSTER_OPERATIONS_ERR_MSG} : {e}"
             Log.error(err_msg)
 
+class NodeMarkFailure(Operation):
```
Comment (pawankumarsrivastava) [inline csm/common/ha/cluster_management/operations.py:184]: NodeMarkFailure --> NodeFailure

Comment (pawankumarsrivastava) [inline csm/common/ha/cluster_management/operations.py:184]: NodeMarkFailure --> NodeFailure

File: csm/common/ha/cluster_management/operations.py [hunk 8]
```diff
@@ -179,3 +181,56 @@ def execute(self, cluster_manager, **kwargs):
...
+    """
+    Process mark node failure request.
+    """
+    def validate_arguments(self, **kwargs):
```
Comment (pawankumarsrivastava) [inline csm/common/ha/cluster_management/operations.py:188]: No need to validate. CSM does not validate semantics. Pass this method

Comment (pawankumarsrivastava) [inline csm/common/ha/cluster_management/operations.py:188]: No need to validate. CSM does not validate semantics

File: csm/common/ha/cluster_management/operations.py [hunk 9]
```diff
@@ -179,3 +181,56 @@ def execute(self, cluster_manager, **kwargs):
...
+        num_node_id = int(Conf.get(const.CSM_GLOBAL_INDEX, const.NUM_NODE_ID))
+        Log.info("Validating node_id")
+        for node_id in range(num_node_id):
+            valid_node_id = Conf.get(const.CSM_GLOBAL_INDEX, f"{const.KEY_NODE_ID}[{node_id}]")
+            if arguments[const.ID] == valid_node_id:
+                return
+        raise InvalidRequest('Request body is missing or invalid request body.')
+
+    def _create_payload(self, **kwargs):
+        """
```
Comment (pawankumarsrivastava) [inline csm/common/ha/cluster_management/operations.py:204]: Remove everything and Pass this one

Comment (pawankumarsrivastava) [inline csm/common/ha/cluster_management/operations.py:204]: Remove everything and Pass this one

File: csm/common/ha/cluster_management/operations.py [hunk 10]
```diff
@@ -179,3 +181,56 @@ def execute(self, cluster_manager, **kwargs):
...
+            f'{HealthAttr.NODE_ID}': node_id,
+            f'{HealthAttr.RESOURCE_TYPE}':const.NODE,
+            f'{HealthAttr.RESOURCE_ID}': node_id,
+            f'{HealthAttr.RESOURCE_STATUS}': const.FAILED,
+            f'{HealthAttr.SPECIFIC_INFO}': ''
+            }
+        return HealthEvent(**health_attrs).json
+
+    def execute(self, cluster_manager, **kwargs):
+        self.validate_arguments(**kwargs)
```
Comment (pawankumarsrivastava) [inline csm/common/ha/cluster_management/operations.py:227]: Remove message bus functionality. Return success for now

Comment (pawankumarsrivastava) [inline csm/common/ha/cluster_management/operations.py:227]: Remove message bus functionality. Return success for now

File: csm/common/ha/cluster_management/operations_factory.py [hunk 11]
```diff
@@ -61,7 +61,8 @@ def get_operation(self, operation: str) -> operations.Operation:
             nodeOperation = operations.NodeStopOperation()
         elif operation == "poweroff":
             nodeOperation = operations.NodePoweroffOperation()
-
+        elif operation == const.MARK_NODE_FAILURE:
+            nodeOperation = operations.NodeMarkFailure()
```
Comment (pawankumarsrivastava) [inline csm/common/ha/cluster_management/operations_factory.py:65]: Change class name

Comment (pawankumarsrivastava) [inline csm/common/ha/cluster_management/operations_factory.py:65]: Change class name

File: csm/conf/csm_default.conf [hunk 12]
```diff
@@ -18,6 +18,11 @@ DEBUG:
   enabled: 'true'
   http_enabled: ''
   default_cluster_id: bae4b468-565d-49df-9495-a43a5d89babc
+SYSTEM:
```
Comment (pawankumarsrivastava) [inline csm/conf/csm_default.conf:21]: No need remove it

File: csm/conf/csm_default.conf [hunk 13]
```diff
@@ -117,6 +122,11 @@ MESSAGEBUS:
         retention_period_ms: '10000'
         method: sync
         partitions: '1'
+    NODE_OPERATIONS:
+      tag_node_failure:
```
Comment (pawankumarsrivastava) [inline csm/conf/csm_default.conf:126]: Remove node failure topics and revert to original. Just keep shutdown signal

Comment (pawankumarsrivastava) [inline csm/conf/csm_default.conf:126]: Remove node failure topics and revert to original. Just keep shutdown signal

File: csm/conf/etc/csm/csm.conf [hunk 14]
```diff
@@ -165,6 +171,11 @@ MESSAGEBUS:
                 retention_period_ms: 10000 # default 10seconds
                 method: "sync"
                 partitions: 1
+        NODE_OPERATIONS:
```
Comment (pawankumarsrivastava) [inline csm/conf/etc/csm/csm.conf:174]: Same as above

Comment (pawankumarsrivastava) [inline csm/conf/etc/csm/csm.conf:174]: Same as above

File: csm/core/blogic/const.py [hunk 15]
```diff
@@ -326,6 +326,8 @@
 PROCESS_CLUSTER_OPERATION_REQ = 'process_cluster_operation'
 PROCESS_CLUSTER_STATUS_REQ = 'process_cluster_status'
 ShUTDOWN_SIGNAL = 'shutdown_signal'
+MARK_NODE_FAILURE = 'mark_node_failure'
+NOT_DEFINED = 'NOT_DEFINED'
```
Comment (pawankumarsrivastava) [inline csm/core/blogic/const.py:330]: Remove it

Comment (pawankumarsrivastava) [inline csm/core/blogic/const.py:330]: Remove it

File: csm/core/blogic/const.py [hunk 16]
```diff
@@ -863,6 +869,9 @@
 MSG_BUS_CLUSTER_STOP_RETENTION_PERIOD = 'MESSAGEBUS>PRODUCER>CLUSTER_MANAGEMENT>cluster_stop>retention_period_ms'
 MSG_BUS_CLUSTER_STOP_PRODUCER_ID = 'MESSAGEBUS>PRODUCER>CLUSTER_MANAGEMENT>cluster_stop>producer_id'
 MSG_BUS_CLUSTER_STOP_METHOD = 'MESSAGEBUS>PRODUCER>CLUSTER_MANAGEMENT>cluster_stop>method'
+MSG_BUS_NODE_FAILURE_MSG_TYPE = 'MESSAGEBUS>PRODUCER>NODE_OPERATIONS>tag_node_failure>message_type'
```
Comment (pawankumarsrivastava) [inline csm/core/blogic/const.py:872]: Remove message bus constants

Comment (pawankumarsrivastava) [inline csm/core/blogic/const.py:872]: Remove message bus constants

File: csm/core/services/cluster_management.py [hunk 17]
```diff
@@ -27,13 +27,6 @@ class ClusterManagementAppService(ApplicationService):
     def __init__(self, plugin, message_bus_obj):
         self._cluster_management_plugin = plugin
         self.message_bus_obj = message_bus_obj
-        self.message_bus_obj.init(type=const.PRODUCER,
```
Comment (pawankumarsrivastava) [inline csm/core/services/cluster_management.py:30]: Revert this back

Comment (pawankumarsrivastava) [inline csm/core/services/cluster_management.py:30]: Revert this back

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]
Rebase it to -> F-24E_Node_Failure

File: csm/common/ha/cluster_management/operations.py (hunk 8 above)
Comment (Pranali04796) [reply csm/common/ha/cluster_management/operations.py:186]: Addressed

Comment (Pranali04796) [reply csm/common/ha/cluster_management/operations.py:186]: resolved

File: csm/conf/csm_default.conf (hunk 12 above)
Comment (Pranali04796) [reply csm/conf/csm_default.conf:21]: keeping it

File: csm/core/services/cluster_management.py (hunk 17 above)
Comment (Pranali04796) [reply csm/core/services/cluster_management.py]: reverted

Comment (Pranali04796) [reply csm/core/services/cluster_management.py]: reverted

File: csm/core/blogic/const.py (hunk 16 above)
Comment (Pranali04796) [reply csm/core/blogic/const.py]: removed message bus related code

Comment (Pranali04796) [reply csm/core/blogic/const.py]: Removed message bus constants

File: csm/core/blogic/const.py (hunk 15 above)
Comment (Pranali04796) [reply csm/core/blogic/const.py]: Done. Removed NOT_DEFINED = 'NOT_DEFINED'

File: csm/conf/etc/csm/csm.conf (hunk 14 above)
Comment (Pranali04796) [reply csm/conf/etc/csm/csm.conf]: removed message bus related configs

Comment (Pranali04796) [reply csm/conf/etc/csm/csm.conf]: removed message bus related configs

**Review by pawankumarsrivastava — APPROVED** [review]

**Pull request merged**